File: mek_stripe_terminal/android/src/main/kotlin/mek/stripeterminal/mappings/SurchargeMappings.kt
@file:OptIn(Surcharging::class)

package mek.stripeterminal.mappings

import com.stripe.stripeterminal.external.Surcharging
import com.stripe.stripeterminal.external.models.SurchargeConfiguration
import com.stripe.stripeterminal.external.models.SurchargeConsent
import com.stripe.stripeterminal.external.models.SurchargeConsentCollection
import com.stripe.stripeterminal.external.models.SurchargeDetails
import com.stripe.stripeterminal.external.models.SurchargeStatus
import mek.stripeterminal.api.SurchargeConfigurationApi
import mek.stripeterminal.api.SurchargeConsentApi
import mek.stripeterminal.api.SurchargeConsentCollectionApi
import mek.stripeterminal.api.SurchargeDetailsApi
import mek.stripeterminal.api.SurchargeStatusApi

fun SurchargeDetails.toApi(): SurchargeDetailsApi {
    return SurchargeDetailsApi(
        amount = amount,
        maximumAmount = maximumAmount,
        status = status?.toApi()
    )
}

fun SurchargeStatus.toApi(): SurchargeStatusApi? {
    return when (this) {
        SurchargeStatus.AVAILABLE -> SurchargeStatusApi.AVAILABLE
        SurchargeStatus.UNAVAILABLE -> SurchargeStatusApi.UNAVAILABLE
    }
}

// PARAMS

fun SurchargeConfigurationApi.toHost(): SurchargeConfiguration {
    val b = SurchargeConfiguration.Builder(amount)
    consent?.let { b.setConsent(it.toHost()) }
    return b.build()
}

fun SurchargeConsentApi.toHost(): SurchargeConsent {
    val b = SurchargeConsent.Builder(collection.toHost())
    notice?.let(b::setNotice)
    return b.build()
}

fun SurchargeConsentCollectionApi.toHost(): SurchargeConsentCollection {
    return when (this) {
        SurchargeConsentCollectionApi.ENABLED -> SurchargeConsentCollection.ENABLED
        SurchargeConsentCollectionApi.DISABLED -> SurchargeConsentCollection.DISABLED
    }
}

File: mek_stripe_terminal/ios/mek_stripe_terminal/Sources/mek_stripe_terminal/Mappings/SurchargeMappings.swift
import Foundation
import StripeTerminal

extension SurchargeDetails {
    func toApi() -> SurchargeDetailsApi {
        return SurchargeDetailsApi(
            amount: amount?.intValue,
            maximumAmount: maximumAmount?.intValue,
            status: status.flatMap { SurchargeStatus(rawValue: UInt($0.intValue)) }?.toApi()
        )
    }
}

extension SurchargeStatus {
    func toApi() -> SurchargeStatusApi? {
        switch self {
        case .available:
            return .available
        case .unavailable:
            return .unavailable
        @unknown default:
            return nil
        }
    }
}

// PARAMS

extension SurchargeConfigurationApi {
    func toHost() throws -> SurchargeConfiguration {
        let b = SurchargeConfigurationBuilder().setAmount(UInt(amount))
        if let it = consent { b.setSurchargeConsent(try it.toHost()) }
        return try b.build()
    }
}

extension SurchargeConsentApi {
    func toHost() throws -> SurchargeConsent {
        let b = SurchargeConsentBuilder().setCollection(collection.toHost())
        if let it = notice { b.setNotice(it) }
        return try b.build()
    }
}

extension SurchargeConsentCollectionApi {
    func toHost() -> SurchargeConsentCollection {
        switch self {
        case .enabled:
            return .enabled
        case .disabled:
            return .disabled
        }
    }
}
